src/PJ_hgridshift.c

```diff
@@ -34,21 +42,36 @@ static LPZ reverse_3d(XYZ xyz, PJ *P) {
     return point.lpz;
 }
 
-
-static PJ_COORD forward_4d (PJ_COORD obs, PJ *P) {
-    obs.xyz = forward_3d (obs.lpz, P);
-    return obs;
+static PJ_COORD forward_4d(PJ_COORD obs, PJ *P) {
+    struct pj_opaque_hgridshift *Q = (struct pj_opaque_hgridshift *) P->opaque;
+    PJ_COORD point = obs;
+    if (Q->t_final != 0.0  && Q->t_epoch != 0.0) {
+        if (obs.lpzt.t < Q->t_epoch && Q->t_final > Q->t_epoch)
```

Comment:
shouldn't that be be <= and >= ?

Or so if t_final and t_epoch are specified, we apply the grid shift only if the observation time is below the epoch, and the epoch below the final epoch. That doesn't immediately seem obvious to me. Should probably be explained in the .rst files

Reply:
I have just updated the .rst files. Do they cover the topic better now?

From a physical point of view I would say that the case where `obs.lpzt.t == Q->t_epoch` is undefined, since the shift doesn't happen instantaneously but rather over a timespan of minutes to days. To me it makes the most sense to say that the offset is applied only if the observation epoch is before the event that caused the deformation. Should you happen to have acquired a coordinate *during* the deformation event it will be partly to fully shifted and we dont' really have the means to do anything about that.

A similar point can be made for the second part of the condition, but I think from a practical point of view it make sense to change it to `Q->t_final >= Q->t_epoch`.

Reply:
Yes the docs make things really clear (sorry I started looking at the code before reading your PR general comment where you mentioned that would be later detailed :-) )
And actually no strong opinion if that should be strict or equal comparisons.